### PR TITLE
Use optimized SmallVec implementation

### PR DESCRIPTION
**Pull request by gootorov**
This PR replaces current SmallVec implementation with the one from the Servo project.

Closes https://github.com/rust-lang/rust/issues/51640

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by gootorov** [comment]
Rebased.

**Comment by Mark-Simulacrum** [comment]
@rust-timer build 31e325e55811a6bde18f86114c56f04729c78575

**Comment by Mark-Simulacrum** [comment]
Perf results look fine to me, will try to get a review in sometime today or tomorrow.

File: src/librustc/Cargo.toml [hunk 1]
```diff
@@ -33,6 +33,7 @@ parking_lot = "0.5.5"
 byteorder = { version = "1.1", features = ["i128"]}
 chalk-engine = { version = "0.7.0", default-features=false }
 rustc_fs_util = { path = "../librustc_fs_util" }
+smallvec = "0.6.4"
```
Comment (llogiq) [inline src/librustc/Cargo.toml]: We may want to use the `union` feature here. (`smallvec = { version = "0.6.4", features = ["union"] }`).

File: src/librustc_allocator/Cargo.toml [hunk 2]
```diff
@@ -16,3 +16,4 @@ rustc_target = { path = "../librustc_target" }
 syntax = { path = "../libsyntax" }
 syntax_pos = { path = "../libsyntax_pos" }
 log = "0.4"
+smallvec = "0.6.4"
```
Comment (llogiq) [inline src/librustc_allocator/Cargo.toml]: Same here.

File: src/librustc_data_structures/Cargo.toml [hunk 3]
```diff
@@ -19,6 +19,7 @@ parking_lot_core = "0.2.8"
 rustc-rayon = "0.1.1"
 rustc-rayon-core = "0.1.1"
 rustc-hash = "1.0.1"
+smallvec = "0.6.4"
```
Comment (llogiq) [inline src/librustc_data_structures/Cargo.toml]: And here.

File: src/librustc_mir/Cargo.toml [hunk 4]
```diff
@@ -25,3 +25,4 @@ syntax = { path = "../libsyntax" }
 syntax_pos = { path = "../libsyntax_pos" }
 byteorder = { version = "1.1", features = ["i128"] }
 rustc_apfloat = { path = "../librustc_apfloat" }
+smallvec = "0.6.4"
```
Comment (llogiq) [inline src/librustc_mir/Cargo.toml]: etc. etc.

**Comment by llogiq** [comment]
I'll push a followup PR to this to change `SmallCStr` to use `SmallVec` once this is merged.

**Comment by Mark-Simulacrum** [comment]
@rust-timer build 40061cbe688a1db38eea0ba1bff9e2a8dc8c8297

File: src/librustc_data_structures/small_vec.rs [hunk 5]
```diff
@@ -234,8 +43,8 @@ mod tests {
         let v: OneVector<isize> = OneVector::new();
         assert_eq!(0, v.len());
```
Comment (Mark-Simulacrum) [inline src/librustc_data_structures/small_vec.rs]: Any of the functionality of smallvec itself shouldn't be tested here I think

**Review by Mark-Simulacrum — COMMENTED** [review]
One nit, but otherwise looks good. Thank you!

File: src/librustc_data_structures/small_vec.rs (hunk 5 above)
Comment (gootorov) [reply src/librustc_data_structures/small_vec.rs]: I guess we sould keep tests for the `expect_one()` function, right? That's the local function that isn't present in smallvec and we implement locally. Also, shall we keep or remove the benchmarks?

Comment (Mark-Simulacrum) [reply src/librustc_data_structures/small_vec.rs]: Remove benchmarks, keep the expect_one test(s).

**Comment by gootorov** [comment]
Addressed review comments.

**Comment by gootorov** [comment]
Rebased and updated `smallvec` to [0.6.5](https://github.com/servo/rust-smallvec/pull/121).

**Comment by michaelwoerister** [comment]
oops... double r+. Looks like it's fine though.

**Pull request merged**